### Load image data

##### Resize one image

In [1]:
import os

In [2]:
from PIL import Image

#### Resizing all the images in all the folders

In [3]:
import pandas as pd
import numpy as np

In [4]:
df = pd.read_csv("training.csv")

In [5]:
df.head()

image image class        date          bar  \
0      training-img/flower/flower_79.jpg      flower  2016-04-05   -13.797746   
1          training-img/boat/boat_10.jpg        boat  2017-12-29  1370.968441   
2  training-img/airplane/airplane_94.jpg    airplane  2016-06-28  1240.615863   
3          training-img/boat/boat_79.jpg        boat  2017-05-31  1258.368516   
4        training-img/llama/llama_25.jpg       llama  2018-03-29   307.458019   

   baz         xgt        qgg          lux       wsg        yyz        drt  \
0    1  132.244709 -14.888019  7059.592259  0.018760 -10.131641  -3.314482   
1    0  -53.010396  -0.249638  5404.074707  0.014199  -0.840672 -57.505186   
2    1  318.103170  25.234516  6435.785062  0.014366  -4.984576  -0.222097   
3    0  221.068821  -5.604373  4915.487849  0.021967   6.189880  -0.732725   
4    0  259.110856   2.896887  5285.831301  0.014830  -7.544632 -22.701896   

         gox        foo       boz  fyt  lgh         hrt         juu  \
0 -19.637941  24.774895  0.077676    1    0 -144.078606  367.977834   
1  -0.477005  43.993301  4.266039    1    1 -169.232629  215.094104   
2  -9.078957   2.444880  1.035487    1    1 -220.349351  220.027026   
3   2.324312  19.290565  0.206948    0    0 -246.892243  495.729658   
4   3.066921  11.171308  0.530466    1    0 -199.510487  234.047119   

          target  
0  134340.064080  
1  167657.624550  
2  157211.302698  
3  -47320.604867  
4  -85209.683794

In [6]:
df.image.head()

0        training-img/flower/flower_79.jpg
1            training-img/boat/boat_10.jpg
2    training-img/airplane/airplane_94.jpg
3            training-img/boat/boat_79.jpg
4          training-img/llama/llama_25.jpg
Name: image, dtype: object

for row in df.image:
    print(row)
    

In [7]:
dfTest = pd.read_csv("test.csv")

In [8]:
dfTest.head()

image        date         bar  baz         xgt  \
0       test-img/boat/boat_3.jpg  2020-11-18  495.361007    0  331.966668   
1  test-img/person/person_53.jpg  2019-01-20  127.415879    1  396.851851   
2      test-img/boat/boat_62.jpg  2019-07-05  -48.353010    0  414.096801   
3         test-img/car/car_7.jpg  2020-12-15   53.163824    0  365.537458   
4         test-img/cat/cat_2.jpg  2020-03-20  -32.591336    0  263.053669   

         qgg          lux       wsg        yyz        drt        gox  \
0  21.021535  5190.710171  0.015209 -12.344110  -4.587213 -11.226993   
1   5.630238  4767.268227  0.010817   5.499534 -25.675941   3.239795   
2  -0.406343  4452.181336  0.010609  -0.746068 -15.213459  -0.518122   
3  15.611367  5864.571557  0.013082 -17.029956 -38.933399  14.905780   
4  -5.938351  4983.384140  0.015802  -7.017901  -7.066405 -25.428807   

         foo       boz  fyt  lgh         hrt         juu  
0  -7.996372  5.512998    1    0 -239.178795  245.806207  
1  23.326502  3.817169    0    1 -192.272913  126.256960  
2  38.568789  3.019702    0    0 -113.590865  121.513203  
3  43.321533  0.935854    1    0  -27.423755  183.329267  
4  37.850267  0.213952    1    1 -192.477269  264.677689

## Find minimum width and height

In [9]:
width, height = [],[]

In [11]:
dir_path = "/Users/Adhira/Downloads/DataScienceTest/training-img/"

In [12]:
subdirs = os.listdir(dir_path)


In [13]:
files = []
for i in range(len(subdirs)):
    if os.path.isdir(dir_path+subdirs[i]):
        file = os.listdir(dir_path+subdirs[i])
#         print(type(file))
        for j in range(len(file)):
#             print(file[j])
            imgPath = os.path.join(dir_path+subdirs[i]+'/'+file[j])
            print(imgPath)
            if os.path.isfile(imgPath):
                img = Image.open(imgPath)
                print(img)
                w,h = img.size
                print(w,h)
                width.append(w)
                height.append(h)

/Users/Adhira/Downloads/DataScienceTest/training-img/cat/cat_15.jpg
<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=157x157 at 0x11379A9B0>
157 157
/Users/Adhira/Downloads/DataScienceTest/training-img/cat/cat_29.jpg
<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=157x157 at 0x113804C50>
157 157
/Users/Adhira/Downloads/DataScienceTest/training-img/cat/cat_44.jpeg
<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=260x260 at 0x113804CC0>
260 260
/Users/Adhira/Downloads/DataScienceTest/training-img/cat/cat_28.jpg
<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=157x157 at 0x113804D30>
157 157
/Users/Adhira/Downloads/DataScienceTest/training-img/cat/cat_16.jpg
<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=157x157 at 0x113804E10>
157 157
/Users/Adhira/Downloads/DataScienceTest/training-img/cat/cat_17.jpg
<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=157x157 at 0x113804E80>
157 157
/Users/Adhira/Downloads/DataScienceTest/training-img/cat/cat_13.jpg
<

<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=157x157 at 0x1138233C8>
157 157
/Users/Adhira/Downloads/DataScienceTest/training-img/dog/dog_55.jpg
<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=157x157 at 0x11380F128>
157 157
/Users/Adhira/Downloads/DataScienceTest/training-img/dog/dog_96.jpg
<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=157x157 at 0x11380F208>
157 157
/Users/Adhira/Downloads/DataScienceTest/training-img/dog/dog_6.jpg
<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=157x157 at 0x11380F278>
157 157
/Users/Adhira/Downloads/DataScienceTest/training-img/dog/dog_32.jpg
<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=157x157 at 0x11380F2E8>
157 157
/Users/Adhira/Downloads/DataScienceTest/training-img/dog/dog_26.jpg
<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=157x157 at 0x11380F358>
157 157
/Users/Adhira/Downloads/DataScienceTest/training-img/dog/dog_33.png
<PIL.PngImagePlugin.PngImageFile image mode=RGBA size=294x500 at 0x1138

<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=157x157 at 0x11378A668>
157 157
/Users/Adhira/Downloads/DataScienceTest/training-img/boat/boat_95.jpg
<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=157x157 at 0x113804E10>
157 157
/Users/Adhira/Downloads/DataScienceTest/training-img/boat/boat_42.jpg
<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=157x157 at 0x113804E80>
157 157
/Users/Adhira/Downloads/DataScienceTest/training-img/boat/boat_56.jpg
<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=157x157 at 0x113804EF0>
157 157
/Users/Adhira/Downloads/DataScienceTest/training-img/boat/boat_68.jpg
<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=157x157 at 0x113804F60>
157 157
/Users/Adhira/Downloads/DataScienceTest/training-img/boat/boat_40.jpg
<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=157x157 at 0x113804FD0>
157 157
/Users/Adhira/Downloads/DataScienceTest/training-img/boat/boat_54.jpg
<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=157

<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=157x157 at 0x113804E10>
157 157
/Users/Adhira/Downloads/DataScienceTest/training-img/llama/llama_4.jpg
<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=157x157 at 0x113804EB8>
157 157
/Users/Adhira/Downloads/DataScienceTest/training-img/llama/llama_5.jpg
<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=157x157 at 0x113804F28>
157 157
/Users/Adhira/Downloads/DataScienceTest/training-img/llama/llama_7.jpg
<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=157x157 at 0x113804F98>
157 157
/Users/Adhira/Downloads/DataScienceTest/training-img/llama/llama_6.jpg
<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=157x157 at 0x113804DA0>
157 157
/Users/Adhira/Downloads/DataScienceTest/training-img/llama/llama_16.jpg
<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=157x157 at 0x113804828>
157 157
/Users/Adhira/Downloads/DataScienceTest/training-img/llama/llama_28.jpg
<PIL.JpegImagePlugin.JpegImageFile image mode=RGB 

In [14]:
minW = min(width)
minH = min(height)

In [15]:
minW, minH

(157, 157)

Minimum size of image will be 157 x 157

## Resizing image

#### New directory

In [16]:
newDir = "/Users/Adhira/Desktop/DataScienceTest/"

In [17]:
for i in range(len(subdirs)):
    if os.path.isdir(dir_path+subdirs[i]):
        file = os.listdir(dir_path+subdirs[i])
#         print(type(file))
        for j in range(len(file)):
#             print(file[j])
            imgPath = os.path.join(dir_path+subdirs[i]+'/'+file[j])
            print(imgPath)
            if os.path.isfile(imgPath):
                img = Image.open(imgPath)
                print(img)
                imNew = img.resize((minW,minH), Image.ANTIALIAS)
                imNew.save(newDir+,inplace=True)

SyntaxError: invalid syntax (<ipython-input-17-84ad38ac423b>, line 13)

### Check sizes

In [18]:
for i in range(len(subdirs)):
    if os.path.isdir(dir_path+subdirs[i]):
        file = os.listdir(dir_path+subdirs[i])
#         print(type(file))
        for j in range(len(file)):
#             print(file[j])
            imgPath = os.path.join(dir_path+subdirs[i]+'/'+file[j])
            print(imgPath)
            if os.path.isfile(imgPath):
                img = Image.open(imgPath)
                print(img.size)

/Users/Adhira/Downloads/DataScienceTest/training-img/cat/cat_15.jpg
(157, 157)
/Users/Adhira/Downloads/DataScienceTest/training-img/cat/cat_29.jpg
(157, 157)
/Users/Adhira/Downloads/DataScienceTest/training-img/cat/cat_44.jpeg
(260, 260)
/Users/Adhira/Downloads/DataScienceTest/training-img/cat/cat_28.jpg
(157, 157)
/Users/Adhira/Downloads/DataScienceTest/training-img/cat/cat_16.jpg
(157, 157)
/Users/Adhira/Downloads/DataScienceTest/training-img/cat/cat_17.jpg
(157, 157)
/Users/Adhira/Downloads/DataScienceTest/training-img/cat/cat_13.jpg
(157, 157)
/Users/Adhira/Downloads/DataScienceTest/training-img/cat/cat_65.jpeg
(260, 260)
/Users/Adhira/Downloads/DataScienceTest/training-img/cat/cat_12.jpg
(157, 157)
/Users/Adhira/Downloads/DataScienceTest/training-img/cat/cat_38.jpg
(157, 157)
/Users/Adhira/Downloads/DataScienceTest/training-img/cat/cat_10.jpg
(157, 157)
/Users/Adhira/Downloads/DataScienceTest/training-img/cat/cat_11.jpg
(157, 157)
/Users/Adhira/Downloads/DataScienceTest/training-i

(157, 157)
/Users/Adhira/Downloads/DataScienceTest/training-img/airplane/airplane_15.jpg
(157, 157)
/Users/Adhira/Downloads/DataScienceTest/training-img/airplane/airplane_29.jpg
(157, 157)
/Users/Adhira/Downloads/DataScienceTest/training-img/airplane/airplane_2.jpg
(157, 157)
/Users/Adhira/Downloads/DataScienceTest/training-img/airplane/airplane_17.jpg
(157, 157)
/Users/Adhira/Downloads/DataScienceTest/training-img/airplane/airplane_16.jpg
(157, 157)
/Users/Adhira/Downloads/DataScienceTest/training-img/airplane/airplane_1.jpg
(157, 157)
/Users/Adhira/Downloads/DataScienceTest/training-img/airplane/airplane_5.jpg
(157, 157)
/Users/Adhira/Downloads/DataScienceTest/training-img/airplane/airplane_12.jpg
(157, 157)
/Users/Adhira/Downloads/DataScienceTest/training-img/airplane/airplane_13.jpg
(157, 157)
/Users/Adhira/Downloads/DataScienceTest/training-img/airplane/airplane_6.jpg
(157, 157)
/Users/Adhira/Downloads/DataScienceTest/training-img/airplane/airplane_38.png
(640, 420)
/Users/Adhira/

In [19]:
from keras.models import Sequential
from scipy.misc import imread
get_ipython().magic('matplotlib inline')
import matplotlib.pyplot as plt
import numpy as np
import keras
from keras.layers import Dense
import pandas as pd

from keras.applications.vgg16 import VGG16
from keras.preprocessing import image
from keras.applications.vgg16 import preprocess_input
import numpy as np
from keras.applications.vgg16 import decode_predictions
train=df.image
test=dfTest.image
# train_path="R/Data/Train/Images/train/"
# test_path="R/Data/Train/Images/test/"

Using TensorFlow backend.


In [20]:
# loading VGG16 model weights
model = VGG16(weights='imagenet', include_top=False)


Instructions for updating:
Colocations handled automatically by placer.


### Getting data from datatset - train

In [21]:
model.summary()

Model: "vgg16"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_1 (InputLayer)         (None, None, None, 3)     0         
_________________________________________________________________
block1_conv1 (Conv2D)        (None, None, None, 64)    1792      
_________________________________________________________________
block1_conv2 (Conv2D)        (None, None, None, 64)    36928     
_________________________________________________________________
block1_pool (MaxPooling2D)   (None, None, None, 64)    0         
_________________________________________________________________
block2_conv1 (Conv2D)        (None, None, None, 128)   73856     
_________________________________________________________________
block2_conv2 (Conv2D)        (None, None, None, 128)   147584    
_________________________________________________________________
block2_pool (MaxPooling2D)   (None, None, None, 128)   0     

In [22]:
dataPath = "/Users/Adhira/Downloads/DataScienceTest/"

In [23]:
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt; 
 
# Importing sklearn libraries
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix, accuracy_score
 
# Importing hypopt library for grid search
from hypopt import GridSearch
 
# Importing Keras libraries
from keras.utils import np_utils
from keras.models import Sequential
from keras.applications import VGG16
from keras.applications import imagenet_utils
from keras.callbacks import ModelCheckpoint
from keras.preprocessing.image import load_img
from keras.preprocessing.image import img_to_array
from keras.layers import Dense, Conv2D, MaxPooling2D
from keras.layers import Dropout, Flatten, GlobalAveragePooling2D
 

/usr/local/lib/python3.7/site-packages/hypopt/model_selection.py:34: UserWarning: If you want to see estimated completion times
    while running methods in cleanlab.pruning, install tqdm
    via "pip install tqdm".
  warnings.warn(w)


In [24]:
def create_features(dataset, pre_model):
    x_scratch = []
    for a in df.image:
        img = Image.open(dataPath+a)
        # load the input image and image is resized to 224x224 pixels
        forResize = img.resize((minW,minH),Image.ANTIALIAS)
        forResize = forResize.convert('RGB')
        forResize.save(dataPath+a)
        image = load_img(dataPath+a, target_size=(minW, minH))
        image = img_to_array(image)
        image = np.expand_dims(image, axis=0)
        image = imagenet_utils.preprocess_input(image)
        x_scratch.append(image)
    x = np.vstack(x_scratch)
    features = pre_model.predict(x, batch_size=32)
    print("features",features)
    print("features shape", features.shape)
    features_flatten = features.reshape((features.shape[0], 4 * 4 * 512))
    return x, features

In [25]:
train_x, train_features = create_features(df, model)

features [[[[ 0.         0.         0.        ...  0.         0.
     0.       ]
   [ 0.         0.         0.        ...  0.         0.
     0.       ]
   [ 0.1162827  0.         0.        ...  0.         0.
     0.       ]
   [ 1.4570134  0.         0.        ...  0.         0.
     0.       ]]

  [[ 0.         0.         0.        ...  0.         0.
     0.       ]
   [ 0.         0.         0.        ...  0.         0.
     0.       ]
   [56.18975    0.         0.        ...  0.         0.
     0.       ]
   [48.227917   0.         0.        ...  0.         0.
     0.       ]]

  [[ 0.         0.         0.        ...  0.         0.
     0.       ]
   [ 0.         0.         0.        ...  0.         0.
     0.       ]
   [ 0.         0.         0.        ...  0.         0.
     0.       ]
   [ 0.         0.         0.        ...  0.         0.
     0.       ]]

  [[ 0.         0.         0.        ...  0.         4.358091
     0.       ]
   [ 0.         0.         0.        ...  0

In [26]:
from keras.layers import Input, Dense, Convolution2D, MaxPooling2D, AveragePooling2D, ZeroPadding2D, Dropout, Flatten, merge, Reshape, Activation

In [28]:
print(train_x.shape)

(678, 157, 157, 3)


In [29]:
train_features_flatten = train_features.reshape((train_features.shape[0], 4 * 4 * 512))

In [30]:
print(train_x.shape, train_features.shape, train_features_flatten.shape)

(678, 157, 157, 3) (678, 4, 4, 512) (678, 8192)


In [31]:
train_x

array([[[[ 1.5106100e+02,  1.3822101e+02,  1.3132001e+02],
         [ 1.5106100e+02,  1.3822101e+02,  1.3132001e+02],
         [ 1.5106100e+02,  1.3822101e+02,  1.3132001e+02],
         ...,
         [ 1.5106100e+02,  1.3822101e+02,  1.3132001e+02],
         [ 1.5106100e+02,  1.3822101e+02,  1.3132001e+02],
         [ 1.5106100e+02,  1.3822101e+02,  1.3132001e+02]],

        [[ 1.5106100e+02,  1.3822101e+02,  1.3132001e+02],
         [ 1.5106100e+02,  1.3822101e+02,  1.3132001e+02],
         [ 1.5106100e+02,  1.3822101e+02,  1.3132001e+02],
         ...,
         [ 1.5106100e+02,  1.3822101e+02,  1.3132001e+02],
         [ 1.5106100e+02,  1.3822101e+02,  1.3132001e+02],
         [ 1.5106100e+02,  1.3822101e+02,  1.3132001e+02]],

        [[ 1.5106100e+02,  1.3822101e+02,  1.3132001e+02],
         [ 1.5106100e+02,  1.3822101e+02,  1.3132001e+02],
         [ 1.5106100e+02,  1.3822101e+02,  1.3132001e+02],
         ...,
         [ 1.5106100e+02,  1.3822101e+02,  1.3132001e+02],
         [

In [32]:
train_features_flatten

array([[0.       , 0.       , 0.       , ..., 0.       , 0.       ,
        0.       ],
       [0.       , 0.       , 0.       , ..., 0.       , 0.       ,
        0.       ],
       [0.       , 0.       , 0.       , ..., 0.       , 0.       ,
        0.       ],
       ...,
       [0.       , 0.       , 0.       , ..., 0.       , 0.       ,
        0.       ],
       [0.       , 0.       , 0.       , ..., 0.       , 0.       ,
        0.       ],
       [0.       , 0.       , 4.2640815, ..., 0.       , 0.       ,
        0.       ]], dtype=float32)

In [33]:
train_features_flatten.shape

(678, 8192)

## Numpy.hstack all features